migration/migration2.go

```diff
@@ -0,0 +1,189 @@
+			go func() {
+				defer func() { <-lim }()
+				var dr deals.StorageDealRecord
+				if err := json.Unmarshal(r.Value, &dr); err != nil {
+					lock.Lock()
+					errors = append(errors, fmt.Sprintf("unmarshaling query result: %s", err))
+					lock.Unlock()
+					return
+				}
+				if dr.DealInfo.DealID != 0 {
+					lock.Lock()
+					ret[dr.DealInfo.ProposalCid] = dr.DealInfo.DealID
+					lock.Unlock()
+				}
+			}()
+		}
+		for i := 0; i < cap(lim); i++ {
+			lim <- struct{}{}
+		}
+
+		if len(errors) > 0 {
+			for _, m := range errors {
+				log.Error(m)
+			}
+			return nil, fmt.Errorf("map build had %d errors", len(errors))
+		}
+	}
+
+	return ret, nil
+}
+
+type v1PartialStorageInfo struct {
+	Cold v1PartialCold
+}
+
+type v1PartialCold struct {
+	Filecoin v1PartialFilecoinInfo
+}
+
+type v1PartialFilecoinInfo struct {
+	Proposals []v1PartialFilStorage
+}
+
+type v1PartialFilStorage struct {
+	ProposalCid cid.Cid
+}
+
+func v2MigrationDealIDFilling(ds datastoreReaderWriter, propsDealID map[cid.Cid]uint64) error {
+	log.Infof("Total map size: %d", len(propsDealID))
+	q := query.Query{Prefix: "/ffs/scheduler/cistore_v2"} // /ffs/scheduler/cistore_v2/<iid>/<cid>
+	res, err := ds.Query(q)
+	if err != nil {
+		return fmt.Errorf("querying storage info store: %s", err)
+	}
+	defer func() { _ = res.Close() }()
+
+	var lock sync.Mutex
+	var errors []string
+	lim := make(chan struct{}, 1000)
+
+	var cantBeMapped, canBeMapped int
+	for r := range res.Next() {
+		if r.Error != nil {
+			return fmt.Errorf("iterating result: %s", r.Error)
+		}
+		lim <- struct{}{}
+
+		r := r
+		go func() {
+			defer func() { <-lim }()
+
+			var oldCi v1PartialStorageInfo
+			var newCi ffs.StorageInfo
+			if err := json.Unmarshal(r.Value, &oldCi); err != nil {
+				lock.Lock()
+				errors = append(errors, fmt.Sprintf("unmarshaling storage info from datastore: %s", err))
+				lock.Unlock()
+				return
+			}
+			if err := json.Unmarshal(r.Value, &newCi); err != nil {
+				lock.Lock()
+				errors = append(errors, fmt.Sprintf("unmarshaling storage info from datastore: %s", err))
+				lock.Unlock()
+				return
+			}
+
+			var filledProposals []ffs.FilStorage
+			for i, p := range oldCi.Cold.Filecoin.Proposals {
+				dealID, ok := propsDealID[p.ProposalCid]
+				if !ok {
+					lock.Lock()
+					cantBeMapped++
+					lock.Unlock()
+					continue
+				}
+				lock.Lock()
+				canBeMapped++
+				lock.Unlock()
+				newCi.Cold.Filecoin.Proposals[i].DealID = dealID
+				filledProposals = append(filledProposals, newCi.Cold.Filecoin.Proposals[i])
+			}
+			newCi.Cold.Filecoin.Proposals = filledProposals
+
+			buf, err := json.Marshal(newCi)
+			if err != nil {
+				lock.Lock()
+				errors = append(errors, fmt.Sprintf("marshaling storage info for datastore: %s", err))
+				lock.Unlock()
+			}
+			if err := ds.Put(datastore.NewKey(r.Key), buf); err != nil {
+				lock.Lock()
+				errors = append(errors, fmt.Sprintf("put storage info in datastore: %s", err))
+				lock.Unlock()
+			}
```

Comment:
Notice that we iterate over `oldCi.Cold.Filecoin.Proposals`, use the _ProposalCid_ from there, and then fill the _DealID_ of **that same entry** but unmarshaled in the current `ffs.StorageInfo`.  Then we save the new filled `ffs.StorageInfo` that now has the _DealID_ information (and indirectly, deleted the old _ProposalCid_ and _ActivationEpoch_ fields since they are not part of the new structure type)

Sidenote: If we happen to not know the mapping, we remove it. We can't leave things with _DealID_ zero.

Reply:
Cool, glad we just discard things we don't have a deal id for. Like I said, I want to know the count, logged below :)